### ISSUE TITLE
Reflect intent to acquire Semaphore synchronously in `Semaphore::acquire`

### ISSUE THREAD

**Issue by greenwoodcm** [tokio-rs/tokio#2675]
**Is your feature request related to a problem? Please describe.**
The `tokio::sync::Semaphore` struct exposes an `acquire()` method that attempts to acquire a permit for the semaphore and will async wait if there are insufficient permits available.  While the `Future` returned by `acquire()` is being awaited, the `Semaphore` has visibility into the fact that there is an outstanding acquirer waiting for the next available permit.  What that means (based on my understanding) is that if in a separate task I invoke `add_permits` to increment the available permit count, I would observe immediately that the permit I just added is not reflected in `available_permits` because it was immediately given back out to the awaiting acquirer.

The problem I face is that the `Semaphore` does not appear to be made aware of the pending acquirer until the returned `Future` is awaited.  This makes complete sense given the way `Semaphore::acquire` is implemented - since it is a plain `async fn` none of the code therein gets executed until the future is awaited.  This makes it difficult to determine from one task if there are other outstanding acquirers of the same `Semaphore` in different tasks.  The current implementation makes it easy to introduce a narrow race condition - where one adds permits, observes that available permits has incremented, and uses that information to infer that there are no outstanding acquirers, when in fact there is an outstanding acquirer (or multiple) that has called `acquire` and just has not awaited the resulting `Future` yet.

**Describe the solution you'd like**
I propose that `Semaphore::acquire` be implemented to synchronously make the `Semaphore` aware of an outstanding acquirer _before_ returning a `Future` to the caller.  That way if `Arc` references are held to the same `Semaphore`, individual tasks can use an add-then-check-available sequence to determine if there are other acquirers out there waiting on the `Semaphore`.

**Comment by Matthias247** [comment]
This contradicts the general contract of `Future`s - which is

> Futures do nothing unless polled

For all places where one would build a composite `Future` first and `.await` or `.poll` it later performing any side effect before polling could lead to some unintended consequences - like blocking the Semaphore too long.

However I'm not sure what the exact problem and impact on semaphores is. Typically one creates the future and immediately `.await`s it. Do you have any particular flow which does not `.acquire().await` and store the intermediate future somewhere, which causes pain?

**Comment by greenwoodcm** [comment]
Here's a condensed example of the code I was trying to write and the problem I ran into:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a266ee63d026630af66abe9961bb8055

In this example I create a `LockFarm` that vends semaphore permits for an arbitrary set of string keys.  i would like acquires on the same key to contend with each other but not acquires of different keys.  So it involves a reasonably straightforward map from key to `Semaphore`.  I take a global lock on the map to insert/fetch a semaphore and then allow contention on that semaphore to proceed while the map lock is freed.  The comment on line 42 is what gave rise to this request for a change in behavior.  On release of a lock I fetch the semaphore out of the map, add the permit back, and immediately check if that permit was taken by an outstanding acquirer.  If not, it _feels_ like I should be able to remove that entry from the map because nobody is waiting to acquire.  But actually, there could be an actor out there with an `Arc` reference to the `Semaphore` for their key that just hasn't awaited their `acquire` future yet.

If the proposed change above were adopted, what I could do is call `acquire` on the semaphore while holding the map lock, thereby guaranteeing that by the time some other actor is attempting to release, the pending acquire is "seen" to the semaphore, will immediately take the next permit, and thus the releaser knows not to remove from the map.

**Comment by greenwoodcm** [comment]
I might propose an extension of the `Semaphore` interface that accomplishes the same effect while adhering to the general contract of futures you quote above.  Obviously pseudo-code, and the naming isn't great, but:

```
struct Semaphore;
impl Semaphore {
    pub fn acquire_slot(&mut self) -> SemaphorePermitSlot {
        // does the synchronous work of putting the acquirer
        // in line so that there is evidence in the Semaphore
        // of an acquirer being present.
    }
}

struct SemaphorePermitSlot;
impl SemaphorePermitSlot {
    pub async fn acquire(mut self) -> SemaphorePermit {
        // translates a permit slot into a permit, doing the
        // actual work of waiting for the semaphore
    }
}
```

**Comment by Darksonn** [comment]
I'm going to close this. I think the current behavior is fine.
